Dynamic Supervisor OIDC Clients proposal (#1126)
vmware-tanzu/pinniped: proposals/1125_dynamic-supervisor-oidc-clients/README.md

  the list of `clientRefs`.

Disadvantages:

- Reusing the example scenario from above (finance and HR clusters), there would be no way to prevent a webapp for
  finance developer users from performing operations against the clusters of HR developer users who log into the finance
  webapp. This could be a serious security problem after the planned multiple identity providers feature is implemented
  to allow for more differentiation between the users of the two FederationDomains. This problem is compounded by the
  fact that many upstream OIDC providers use browser cookies to avoiding asking an active user to interactively log in
  again, and also by the fact that we decided to punt implementing a user consent UI screen. Together, these imply that
  an attacker from the finance department cluster which runs the client webapp would only need to trick an HR user into
  clicking on a single link in their web browser or email client for the attacker to be able to gain access to the HR
  clusters using the identity of the HR user, with no further interaction required by the HR user beyond just clicking
  on the link.

> 

Reviewer: Isn't there a third option where the association is configured on the `OIDCClient`? Since only admins can configure clients, this should be as safe as configuring it on the `FederationDomain`. It  has the advantages of not requiring a `FederationDomain` update when you create a new client, and not requiring implementation work on our part to change the federation domain controller.

Author: Yes, that would also be possible.

The `FederationDomain` is the configuration of the downstream OIDC issuer, so in my mind it would make sense for the issuer to decide for itself which clients are allowed on the issuer. The burden on the user is about the same, i.e. when you create or delete a new issuer then you don't need to update any `OIDCClients` configuration. I'd expect us to do the same thing when we implement multiple IDPs, which is list the allowed IDPs on the `FederationDomain`, so that would be consistent if we use the same approach here. (In the case of IDPs it would also allow us to put extra per-`FederationDomain` configuration on each IDP at that "join" point, but I don't think that applies here for `OIDCClients`.)

But either alternative would be possible as an API.

CLIs with old Supervisors, in regard to the new Supervisor features proposed herein.

#### Tests

As usual, unit tests should be added for all new/changed code.

Integration tests should be added to mimic the usage patterns of a webapp. Dynamic clients should be configured with
various options to ensure that the options work as expected. Those clients should be used to perform the authcode flow,
RFC8693 token exchanges, and TokenCredentialRequest calls. Negative tests should include validation failures on the new
CRs, and failures to perform actions that are supposed to be disallowed on the client by its configuration.

#### New Dependencies

None are foreseen.

#### Performance Considerations

Some considerations were mentioned previously for client secret option 2 above. No other performance impact is foreseen.

#### Observability Considerations

None are foreseen, aside from the usual error messages and log statements for the new features similar to what is
already implemented in the Supervisor.

#### Security Considerations

Some security considerations were already mentioned above. Here are a couple more.

##### Ensuring reasonable client secrets

> During a login flow, when the client secret is presented in plaintext to the Supervisor’s token endpoint, it could

Reviewer: Would this validation print some warning to the logs, or prevent logins?

Author: I would expect it to prevent logins in order to get the attention of the admin. It could also put a nice message in the logs to help the admin debug.

Reviewer: Makes sense to me.

  password, which should only be seen by the Supervisor and the LDAP server. However, the details of this item have been
  split out to a separate proposal document.
- Client secrets should be stored encrypted or hashed, not in plain text.

Non-goals for this proposal:

- Pinniped's scope is to provide authentication for cluster users. Providing authentication for arbitrary users to
  arbitrary webapps is out of scope. The only proposed use case is providing the exact same identities that are provided
  by using Pinniped's `kubectl` integration, which are the developers/devops/admin users of the cluster.
- Supporting any OAuth/OIDC flow other
  than [OIDC authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).
- Implementing any self-service client registration API. Clients will be registered by the Pinniped admin user.
- Implementing a consent screen. This would be clearly valuable but will be left as a potential future enhancement in
  the interest of keeping the first draft of this feature smaller.
- Management (ie creation & rotation) of client credentials on the operator's behalf. This will be the operator's

>   responsibility.

Reviewer 2: Agreed, this non-goal should be updated!

  the status.

Some other settings are implied and will not be configurable:

- All clients must use [PKCE](https://oauth.net/2/pkce/) during the authorization code flow. There is a risk that some
  client libraries might not support PKCE, but it is considered a modern best practice for OIDC.
- All clients must use [client secret basic auth](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1) for
  authentication at the token endpoint. This is the most widely supported authentication method in client libraries and
  is recommended by the OAuth 2.0 spec over the alternative of using query parameters in a POST body.
- All clients are only allowed
  to [use `code` as the `response_type`](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationExamples)
  at the authorization endpoint.
- All clients are only allowed to use the default or specify `query` as
  the [`response_mode`](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) at the authorization
  endpoint. This excludes clients from using `form_post`. We could consider allowing `form_post` in the future if it is
  desired.

> - Clients are not allowed to use JWT-based client auth. This could potentially be added as a feature in the future.

Reviewer 2: I dunno that this is a "non goal" its just not likely going to be the option that's chosen for the MVP.

Regardless, of how that decision communicated in this proposal, let's be sure to create an issue for implementing [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523) support in the future.

assumption. For example, deployment tools will avoid showing the values in Secrets during an application deployment. As
another example,
[Kubernetes best practices suggest](https://kubernetes.io/docs/concepts/configuration/secret/#information-security-for-secrets)
that admins should use authorization policies to restrict read permission to Secrets as much as possible. Additionally,
some clusters may use the Kubernetes feature to
[encrypt Secrets at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/), and thus reasonably expect
that all confidential data is encrypted at rest.

###### Option 1: Providing client secrets already hashed

An admin could run bcrypt themselves to hash their desired client secret. Then they could write the resulting value into
a Secret.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: my-webapp-client-secret-1
  namespace: pinniped-supervisor
type: secrets.pinniped.dev/oidc-client-secret-bcrypt
stringData:
  clientSecret: $2y$10$20UQo9UzqBzT.mgDRu9TwOC...EQSbS2
```

Advantages:

- Least implementation effort.
- Admins choose their own preferred bcrypt input cost.
- Confidential data is stored in a Secret.

> Disadvantages:

Reviewer 2: If the aggregated API route is taken for the OIDCClient API, then client names and secrets could both be generated, correct?

> Disadvantages:

Reviewer 3: > If the aggregated API route is taken for the OIDCClient API, then client names and secrets could both be generated, correct?

Yes to both.  I modeled the secrets bit after existing Kube APIs to match their UX.  Re: names, we would probably need to split up the `metadata.name` from the generated client ID (i.e. in the same way that the secret is split up) so that you could consistently reference a particular Kube object even if you lack control over its client ID.

FederationDomain for all the clusters used by the HR department, then a webapp for the finance department developers
should not necessarily be allowed to perform actions on the Kubernetes API of the HR department's clusters.

###### Option 1: Explicitly associate specific clients with issuers

Each FederationDomain could list which clients are allowed. For example:

```yaml
apiVersion: config.supervisor.pinniped.dev/v1alpha1
kind: FederationDomain
metadata:
  namespace: pinniped-supervisor
  name: my-domain
spec:
  issuer: https://my-issuer.pinniped.dev/issuer
  # This is the new part...
  clientRefs:
    - "my-webapp-client"
    - "my-other-webapp-client"
```

The `pinniped-cli` client does not need to be listed, since it only makes sense to allow `kubectl` access to all users
of all FederationDomains. Additionally, the `pinniped-cli` can only redirect authcodes to localhost listeners,
effectively only allowing users to log into their own accounts on their own computers.

###### Option 2: Implicitly associate all clients with all issuers

Rather than explicitly listing which clients are allowed on a FederationDomain, all FederationDomains could assume that

> all clients are available for use.

Reviewer 2: +1 I agree, I think its desirable for the dynamic clients to have the same relationship with FederationDomains as the CLI client. 

As mentioned elsewhere in the comments on this proposal, it is not possible to meaningfully enable multiple FederationDomain's today.  This modeling decision can revisited if/when Pinniped is modified to allow that outcome.

   We could consider inventing a way to make this initial ID token more inert. One possibility would be to take
   advantage of the
   [`RequiredClaims` field](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L117-L119)
   of the JWT authenticator. The token exchange could be enhanced to always add a new custom claim to the JWTs that it
   returns, such as `pinniped_allow_concierge_tcr: true`. The Concierge JWTAuthenticator could be enhanced to require
   this claim by default, along with a configuration option to disable that requirement for users who are not using the
   Supervisor. ID tokens returned during an initial login (authcode flow) would not include this claim, rendering them
   unusable at the Concierge's TokenCredentialRequest endpoint. Docs could be updated to explain that users who
   configure dynamic clients should upgrade to use the version of the Concierge which performs this new validation on
   workload clusters, and that users who are using JWTAuthenticators for providers other than the Supervisor would need

>    to add config to disable the new validation when they upgrade.

Author: @margocrawf @enj Is there a better way to handle this case perhaps? I think this concept would work but it's sort of a breaking change for a minority of users, so maybe there's a better way. Any ideas?

Reviewer: It would be neat if we could do the opposite (e.g. some claim that only exists in the ID token returned during the initial login, but not included in the exchanged tokens or in ID tokens from somewhere other than the supervisor).
Then we could check for the absence of that claim.
But it seems like that would be more work since it's not supported the way `RequiredClaims` is.

## Terminology / Concepts

- See the [definition of a "client" in the OAuth 2.0 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-1.1).
  For the purposes of this proposal, a "client" is roughly equal to a webapp which wants to know the authenticated
  identity of a user, and may want to perform actions as that user on clusters. An admin needs to allow the client to
  learn about the identity of the users by registering the client with the Pinniped Supervisor.
- See also the [OIDC terminology in the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#Terminology).
- The OIDC clients proposed in this document are "dynamic" in the sense that they can be configured and reconfigured on
  a running Supervisor by the admin.

## Proposal

### Goals and Non-goals

Goals for this proposal:

- Allow Pinniped admins to configure applications (OIDC clients) other than the Pinniped CLI to interact with the
  Supervisor.
- Provide a mechanism which governs a client's access to the token exchange APIs. Not all webapps should have permission
  to act on behalf of the user with the Kubernetes API of the clusters, so an admin must be able to configure which

>   clients have this permission.

Author: @absoludity Sorry for not being more clear here. There are two token exchange APIs.

The first is the OIDC token endpoint of the Supervisor. It can be used to exchange an OIDC access token for a new cluster-scoped OIDC ID token (with the audience changed to be the audience name of the workload cluster). That's the exchange to which this bullet point refers. An OIDC may or may not be given permission to perform this kind of exchange.

The other is the TokenCredentialRequest of the Concierge that you mention. That endpoint is available to anon users as you pointed out. This proposal does not suggest any changes to that API.

>   clients have this permission.

Reviewer 4: Thanks - right, I've not ever used supervisor as we weren't able to provide a different client (other issue, which may be fixed now), so much of my commenting here may have been in the wrong context, sorry.

- Provide a mechanism for requesting access to different aspects of a user identity, especially getting group

>   memberships or not, to allow the admin to exclude this potentially information for clients which do not need it.

Author: Yes, if the client is allowed tp request groups, and if it does actually request groups, then the resulting ID token will contain the users groups, and therefore the resulting k8s mTLS client cert credential will also contain the user's groups. This would be the "normal" use case.

There is another potential use case for an OIDC client. It might represent a web app which wants to know that the user has authenticated, and maybe wants to know the user's username, but does not need to know the user's groups and does not need to make k8s API calls on behalf of the user. That type of webapp would not need the groups scope.

- Support a web UI based LDAP/ActiveDirectory login screen. This is needed to avoid having webapps handle the user's
  password, which must only be seen by the Supervisor and the LDAP server. However, the details of this item have been

>   split out to a separate proposal document.

Reviewer 4: Yes, that's exactly what I meant. And yes, we currently setup our dev environment with Dex + openldap to ensure ldap works. If we switched to use supervisor, I can do that too, thanks.

- Client secrets must be stored encrypted or hashed, not in plain text.
- Creation of client credentials on the operator's behalf - the server must generate any secrets.
- The operator must be able to initiate manual rotation of client credentials.

> - Documentation describing the token exchanges a webapp backend must perform to interact with the Kubernetes API.

Author: Yes, we're considering such a feature, but it was not included in the scope of this proposal to keep the scope smaller. It might be considered as a follow-up proposal. Is that a feature that you would like to see? (cc @anjaltelang)

> - Documentation describing the token exchanges a webapp backend must perform to interact with the Kubernetes API.

Reviewer 4: Yes - I think if that feature was provided, we could completely remove our `pinniped-proxy` service (as much as I've enjoyed creating and working on a small Rust project :) )

  by using Pinniped's `kubectl` integration, which are the developers/devops/admin users of the cluster.
- Supporting any OAuth/OIDC flow other
  than [OIDC authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).
- Implementing any self-service client registration API. Clients will be registered by the Pinniped admin user.
- Implementing a consent screen. This would be clearly valuable but will be left as a potential future enhancement in
  the interest of keeping the first draft of this feature smaller.
- Orchestration of token exchanges on behalf of the client. Webapps which want to make calls to the Kubernetes API of
  clusters acting as the authenticated user will need to perform the rest of the token and credential exchange flow that
  it currently implemented by the Pinniped CLI. Providing some kind of component or library to assist webapp developers
  with these steps might be valuable but will be left as a potential future enhancement.
- Supporting JWT-based client auth as described in [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523).  For now

>   client secret basic auth will be used.  This is left as a potential future enhancement.

Author: That's right. This bullet point is only referring to how a client authenticates during the authcode flow. We won't support JWT-based authentication of the OIDC client during the authcode flow (yet).

> - Supporting public clients.

Author: In this context, "public" clients refers to clients that are not required to authenticate in authcode flows. E.g. the Pinniped CLI uses a public OIDC client, since there would be no point in hardcoding a client secret into the CLI client code, thus rendering it a non-secret.

> - Supporting public clients.

Reviewer 4: OIC, yes, that was not clear to me at all, but makes complete sense with the extra context.

- Supporting any policy around which users an OAuth client can interact with.  This is left as a potential future
  enhancement.

### Specification / How it Solves the Use Cases

This document proposes supporting a new Custom Resource Definition (CRD) for the Pinniped Supervisor which allows the
admin to create, update, and delete OIDC clients for the Supervisor.

#### API Changes

##### Configuring clients

An example of the new CRD to define a client:

```yaml
apiVersion: oauth.supervisor.pinniped.dev/v1alpha1
kind: OIDCClient
metadata:
  name: client.oauth.pinniped.dev-my-webapp-client

>   namespace: pinniped-supervisor

Reviewer 4: Do users need to be able to create the `OIDCClient` in the `pinniped-supervisor` namespace, or can they be created in the namespace in which the user has the relevant RBAC (EDIT: I see below that you've outlined it has to be in the same namespace as the supervisor as it is configuration for the supervisor)?

Reviewer 4: Well, might be harder to control delegated permissions from having read/write access on each other's (without something more granular like open policy agent) - which is why I would have thought namespaces could be useful if there was any thought of delegation. But ok, if it's admin-only, makes sense.

`.spec.generateNewSecret` set to `true` to cause the creation of a new client secret.  The response will include the
total number of client secrets (including any newly generated ones) that exist for the OAuth client in the
`.status.totalClientSecrets` field.  When the admin is ready, they may call the API with `.spec.revokeOldSecrets` set to
`true` to cause all but the latest secret to be revoked.  In the event of a client secret disclosure, a "hard" rotation
may be performed by setting both `.spec.generateNewSecret` and `.spec.revokeOldSecrets` to `true` (this will revoke all
pre-existing client secrets and return a newly generated secret).  Leaving both of these fields set to `false` will
simply return the number of existing client secrets via the `.status.totalClientSecrets` field (this same information is
available via the `.status.totalClientSecrets` field of the `OIDCClient` resource).

An admin would interact with this API by using standard `kubectl` commands:

```yaml
apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
kind: OIDCClientSecretRequest
metadata:
  name: client.oauth.pinniped.dev-j77kz

>   namespace: pinniped-supervisor

Reviewer 4: So if I, as a user, need RBAC to `create` an `OIDCClintSecretRequest` in the `pinniped-supervisor` namespace to be able to create new secrets or revoke old ones for my client, is there some finer-grained permissions required to stop me from doing so with other clients? If it's that only an admin with access to all clients can do this (ie. effectively a cluster-admin), then how am I safely passing the resulting client secret to the team that will use it with their web app or similar? I'm wondering if these requests should instead be in my own namespace so that as an admin of my namespace I can get the client secret etc.?

Author: That's a good question. These clients are not meant to be self-serve created by the web app developer (yet). They are meant to be managed by the Supervisor admin, so the Supervisor admin would register a client for your app and then send you the credential. Supporting the self-service model is left for future possible consideration.

Reviewer 4: Yeah ok. Still seems to me like the request belongs in the users namespace given that's where the secret would usually end up anyway (I think?), even if it requires an admin to approve by creating some related resource (for now) in the supervisor namespace. The controller could even, once the admin has manually approved the request somehow, create the *secret* resource in the same (user) namespace so that the disappearing status secret below isn't an issue (assuming the secret is going to end up being a secret resource in the user namespace that gets mounted somewhere anyway?). But I'm probably lacking lots of context here, could be many other reasons for the current approach. And perhaps the client secret won't need to be a secret resource for some users.

spec:
  generateNewSecret: true
```

Assuming the above yaml is stored in `file.yaml`, then running:

`kubectl create -f file.yaml`

would cause the server to respond with (note the custom columns in the table output):

```
NAMESPACE               NAME                                SECRET     TOTAL
pinniped-supervisor     client.oauth.pinniped.dev-j77kz     12..xz     2

> ```

Reviewer 4: Won't I then miss the generated secret in the response status here unless I requested `-o yaml` or similar?

Author: @enj ?

Reviewer 4: Related thought (based on above): iff the generated secret would usually end up being manually added as a secret resource in the users namespace, why not have the controller create the secret resource for the user rather than relying on catching the status in the first response. I think the same semantics for revocation etc. could still apply. It just depends whether the assumption can be made that a secret in the user's (requested?) namespace is secure.

Author: The user needs the plaintext secret so they can put that in their webapp's client config. @enj wanted to avoid storing the client secret in plain text anywhere, and also wanted the Supervisor to auto-generate the client secret to ensure that it is a strong (long, random) password. These two requirements together imply that the Supervisor needs to return the plaintext password one time upon creation, and then store the hashed version. Unfortunately, these requirements add a lot of complexity to both the design and implementation, but is the direction that we're going at the moment.

Reviewer 4: > The user needs the plaintext secret so they can put that in their webapp's client config.

Yep, I guess I'm just asking if that client config will always end up being a combination of a k8s config map and k8s secret anyway (ie. will the user be creating a k8s secret with the generated client-secret to use in their webapp's config). If not, then I totally agree with the first requirement (as well as the second of course).

Reviewer 3: There is no requirement that the app using the dynamic client be hosted on the same cluster as the supervisor (and I think we would prefer if it was not).  Said app may not even be running on Kubernetes.  Even when running on Kube, https://secrets-store-csi-driver.sigs.k8s.io can be used to store the client secret in an external vault and mounted into the app's pod without using Kubernetes secrets (which are really bad at actually being secret).

Reviewer 3: > Won't I then miss the generated secret in the response status here unless I requested `-o yaml` or similar?

Sadly yes.  Even though all the code exists in the API server to return a proper table output on `create` (implemented in #1235), `kubectl` does not request it 😭

supervisor will issue warnings and request that the user contact their admin for an updated kubeconfig.  This exclusion
will be dropped in a future release.  Having this temporary exclusion will allow mixing of old CLIs with new
supervisors, and new CLIs with old supervisors.  New CLIs will automatically include these new scopes in the kubeconfigs
that they generate.

#### Tests

As usual, unit tests must be added for all new/changed code.

Integration tests must be added to mimic the usage patterns of a webapp. Dynamic clients must be configured with
various options to ensure that the options work as expected. Those clients must be used to perform the authcode flow,
RFC8693 token exchanges, and TokenCredentialRequest calls. Negative tests must include validation failures on the new
CRs, and failures to perform actions that are supposed to be disallowed on the client by its configuration. Integration
tests will be used to mimic an upgrade scenario (with and without dynamic clients) to confirm that audience confusion is
not possible.

#### New Dependencies

None are foreseen.

#### Performance Considerations

> Extra calls will be made to the Kubernetes API to lookup dynamic OAuth clients.  No performance impact is foreseen.

Reviewer 4: You mentioned the bcrypt performance consideration above, might be worth including here?

Author: Good point, thanks!

Author: Added.

      `allowedScopes` lists `pinniped:request-audience`.
- `allowedScopes`: Decide what the client is allowed to request. Note that the client must also actually request
  particular scopes during the authorization flow for the scopes to be granted. May only contain the following valid
  options:
    - `openid`: The client is allowed to request ID tokens.  ID tokens only include the
      [required claims](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) by default (`iss`, `sub`, `aud`,
      `exp`, `iat`).  This scope must always be listed.
    - `offline_access`: The client is allowed to request an initial refresh token during the authorization code grant
      flow.  This scope must be listed if `allowedGrantTypes` lists `refresh_token`.
    - `pinniped:request-audience`: The client is allowed to request a new audience value during a RFC8693 token
      exchange, which is a step in the process to be able to get a cluster credential for the user.  `openid`,
      `username` and `groups` scopes must be listed when this scope is present.  This scope must be listed if
      `allowedGrantTypes` lists `urn:ietf:params:oauth:grant-type:token-exchange`.

>     - `username`: The client is allowed to request that ID tokens contain the user's username.  This is a newly

Reviewer: I'm not sure that the `username` scope is necessary. I think it's acceptable to always include the username.

Reviewer 3: > I'm not sure that the `username` scope is necessary. I think it's acceptable to always include the username.

The same reasoning could be applied to groups.  I think it is better to be specific about exactly what information is shared and when.  For example, in the future if we supported an `email` claim, a web app could be granted access to that claim but not the username.

resources. This makes it explicit that the data is confidential.
[Kubernetes best practices suggest](https://kubernetes.io/docs/concepts/configuration/secret/#information-security-for-secrets)
that admins should use authorization policies to restrict read permission to Secrets as much as possible. Additionally,
some clusters may use the Kubernetes feature to
[encrypt Secrets at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/), and thus reasonably expect
that all confidential data is encrypted at rest.  We will use Kubernetes secrets to store the client secret hash.

CRDs are convenient to use, however, they have one core limitation - they cannot represent non-CRUD semantics.  For
example, the existing token credential request API could not be implemented using a CRD - it processes an incoming token
and returns a client certificate without writing any data to etcd.  Aggregated APIs have no such limitation.  We will
use an aggregated API to handle generation of client secrets.

We will use the existing token credential request API as a model for the new OIDC client secret request API.  Only the

> `create` verb will be supported (but this resource will be part of the `pinniped` category and will have no-op `list`

Reviewer: While I think this may be fine for an mvp, I also think being able to use `kubectl delete` would make the experience a lot nicer and more intuitive for users who are used to kubectl based apis and I think we should prioritize that work soon.

Reviewer 3: Agree, and nothing prevents us from building that stuff in the future.  There is also some value and waiting for feedback.

   traditionally done with Dex.

2. A user can use the public `pinniped-cli` client to log in and then to perform a token exchange to any audience value.
   They could even ask for an audience which matches the name of an existing dynamic client to get back an ID token that
   would appear as if it were issued to that dynamic client. Then they could try to find a way to use that ID token with
   a webapp which uses that dynamic client to authenticate its users (although that may not be possible depending on the
   implementation of the webapp).

To address all of these issues we will:

- Require a static, reserved prefix (`client.oauth.pinniped.dev-`) for all dynamic OAuth client IDs
- Require a static, reserved prefix (`token-exchange.oauth.pinniped.dev-`) for audiences requested via the token
  exchange API
- All ID tokens issued by the supervisor will contain the `azp` claim and its value will always be set to the client ID
  of the client that was used for the initial login flow.  This is meant to prevent any information loss during flows
  such as the token exchange API.

> - In the ID token retrieved from the authorization flow (and only this ID token), nest the `username` and `groups`

Reviewer 3: > I think it will make our code [more confusing]

The code change will be trivial so I do not consider this as a valid argument against.

> and our tokens more confusing.

They will certainly look weird but passport JS and oauth2-proxy can easily consume nested claims.  The alternative is to shove all the claims behind a user info endpoint, which is more work to implement for us and maybe more work for a web app to consume.

> My preference would be to be as clear as we can in docs but not impose any hard limitation.

Just as we are deprecating and removing the http port to impose a hard limitation on your network setup, I believe we should do the same with how you configure and consume our auth APIs.  Insecure configs should simply not be possible.

> - In the ID token retrieved from the authorization flow (and only this ID token), nest the `username` and `groups`

Reviewer 3: Thought: what breaks if only the initial authorization code flow returns an ID token (i.e. it is not returned on refresh flows)?  We would be OIDC spec compliant but incompatible with the `kubectl` + API server + Dex + Gangway style approach.  I _think_ the only thing that would be missing for a web app is group membership updates?

      `username` and `groups` scopes must be listed when this scope is present.  This scope must be listed if
      `allowedGrantTypes` lists `urn:ietf:params:oauth:grant-type:token-exchange`.
    - `username`: The client is allowed to request that ID tokens contain the user's username.  This is a newly
      proposed scope which does not currently exist in the Supervisor. Without the `username` scope being requested and
      allowed, the ID token would not contain the user's username.
    - `groups`: The client is allowed to request that ID tokens contain the user's group membership, if their group
      membership is discoverable by the Supervisor. This is a newly proposed scope which does not currently exist in the
      Supervisor. Without the `groups` scope being requested and allowed, the ID token would not contain groups.
- `phase`: This enum (`Pending`,`Ready`,`Error`) summarizes the overall status of the client (defaults to `Pending`).
- `totalClientSecrets`: The number of client secrets that are currently associated with this client.
- `conditions`: The result of validations performed by a controller on these CRs will be written by the controller on

>   the status. And example condition is showed above, and other conditions will also be added by the controller.

Reviewer 3: ```suggestion
  the status. An example condition is showed above, and other conditions will also be added by the controller.
```

Author: Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.

A dynamic client with many client secrets will be slower to authenticate (especially slow when failing to authenticate),
since each hashed secret must be tried in sequence until one works, and the cost of trying each is relatively high.
Administrators should take care to remember to come back to use this endpoint again to revoke the old client secrets
once they are not needed anymore.

An admin would interact with this API by using standard `kubectl` commands:

```yaml
apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
kind: OIDCClientSecretRequest
metadata:
  name: client.oauth.pinniped.dev-j77kz
  namespace: pinniped-supervisor
spec:
  generateNewSecret: true
```

Assuming the above yaml is stored in `file.yaml`, then running:

`kubectl create -f file.yaml`

would cause the server to respond with (note the custom columns in the table output):

```
NAMESPACE               NAME                                SECRET     TOTAL
pinniped-supervisor     client.oauth.pinniped.dev-j77kz     12..xz     2
```

All client secret hashes for an OAuth client will be stored in `pinniped-storage-oidc-client-secret-<metadata.uid>` where

> `metadata.uid` is a hashed version of the UID generated by the Kuberentes API server for the `OIDCClient` custom resource instance that

Reviewer 3: ```suggestion
`metadata.uid` is a base32 encoded version of the UID generated by the Kuberentes API server for the `OIDCClient` custom resource instance that
```

Author: Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.

represents the given OAuth client.  This secret will have an owner reference back to the `OIDCClient` to guarantee that
it is automatically garbage collected if the `OIDCClient` is deleted.  As the client secret lookup is UID based, it is
resilient against any vulnerabilities that arise from the client ID being re-used over time.  Using the UID also
prevents any issues that could arise from giving the user control over the secret name (i.e. length issues, collision
issues, validation issues, etc).  Each client will have a 1:1 mapping with a Kubernetes secret - there would always be
at most one Kubernetes secret per client.  Since the secret name is deterministic based on the client UID, no reference
field is required on the client.  Kubernetes secrets have a size limit of ~1 MB which is enough to hold many thousands

> of hashes.  This API will enforce a hard limit of 100 secrets per client (having this many client secrets is likely a

Reviewer 3: ```suggestion
of hashes.  This API will enforce a hard limit of 5 secrets per client (having this many client secrets is likely a
```

This is what I ended up doing in #1235 to avoid super slow authentications.

Author: Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.

  use. Any token exchange requesting an audience with that substring will be rejected. This leaves room to create other
  categories of audience values in case that is needed in the future, e.g.
  `static-client.oauth.pinniped.dev-some-client`.
- All ID tokens issued by the supervisor will contain the `azp` claim and its value will always be set to the client ID
  of the client that was used for the initial login flow.  This is meant to prevent any information loss during flows
  such as the token exchange API.
- No changes are proposed for the `JWTAuthenticator` as it is meant to be interchangeable with the Kubernetes OIDC
  server flags.  Some environments use old versions of the concierge with newer versions of the supervisor and thus we
  cannot rely on changes to the concierge being rolled out to enforce security contracts.
- Implicit behavioral changes to APIs are avoided as they can be difficult to understand and reason about.

The validation associated with dynamic clients will be used to enforce that clients have `metadata.name` set to a value
that starts with `client.oauth.pinniped.dev-`.  Thus while it will be possible to create a dynamic client CR with a name

> such as `pinniped-cli`, it will not be valid and will not be usable.  For the time being, the hardcoded `pinniped-cli`

Reviewer 3: I think the CRD regex prevents creation of such clients.

Author: Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.

- Management (ie creation & rotation) of client credentials on the operator's behalf. This will be the operator's
  responsibility.
- Orchestration of token exchanges on behalf of the client. Webapps which want to make calls to the Kubernetes API of
  clusters acting as the authenticated user will need to perform the rest of the token and credential exchange flow that
  it currently implemented by the Pinniped CLI. Providing some kind of component or library to assist webapp developers
  with these steps might be valuable but will be left as a potential future enhancement.

### Specification / How it Solves the Use Cases

This document proposes supporting a new Custom Resource Definition (CRD) for the Pinniped Supervisor which allows the
admin to create, update, and delete OIDC clients for the Supervisor.

#### API Changes

##### Configuring clients

An example of the new CRD to define a client:

```yaml
apiVersion: oauth.supervisor.pinniped.dev/v1alpha1
kind: OIDCClient
metadata:
  name: my-webapp-client
  namespace: pinniped-supervisor
spec:
  id: my-webapp
  secretNames:
    - my-webapp-client-secret-1
    - my-webapp-client-secret-2
  allowedRedirectURIs:

>     - https://my-webapp.example.com/callback

Reviewer 3: Perhaps we should limit this to a single string instead of a list to discourage the use of the same client across multiple web apps.

Reviewer 4: That would be different to most other client definitions, wouldn't it? (not that it's a reason not to do so)

Reviewer 4: > @absoludity I'm not sure why, but GitHub won't let me reply to your comment which said "That would be different to most other client definitions, wouldn't it? (not that it's a reason not to do so)".
> 
> I'm not sure what you mean by that comment. Could you explain? Thanks!

I meant simply that other providers that I've used have always allowed multiple redirect URIs (though might just be the ones I've used). It'd be a surprise to find I couldn't do so, but may not be a bad thing :) (My main reason for using multiple redirect URLs is usually because I've got a client setup for use in different development projects, so not a good reason to allow it here of course).

Author: Ah, yes, thanks for explaining. We're planning to allow multiple redirect URIs to be configured.

Some key differences from the CRD based approach:

- The `.status` field is omitted - validation is performed statically on every `create` and `update` operation - the
  client object is always guaranteed to be valid.  The admin is notified directly via a validation error if they attempt
  to create an incorrect object (for example, with `metadata.name` set to a value that includes a `:`).  Similarly, if
  the `.spec.allowed*` fields include invalid data, the object will be rejected.  Consumers of this API, such as our
  controllers, can rely on the data being valid, which simplifies their logic.
- `.spec.secretNames` has been removed as the client secret is generated on the server after the client has already been
  created.  This is handled via a `create` operation to the `secret` subresource (this is modeled after the Kubernetes
  service account API's `token` subresource).  Just as the SA `token` subresource uses the `TokenRequest` API as the
  input and output schema, the `secret` subresource would use a `SecretRequest` API as the input and output schema.
  The response from this API is the only time that the plaintext client secret is made available.  Since rotation is not

>   part of this proposal, the `secret` API can only be invoked once per client - all subsequent requests will fail.  A

Reviewer 2: I don't think "rotation is not a part of this proposal" is an accurate representation of the non goal

> Management (ie creation & rotation) of client credentials on the operator's behalf. 

The intention of that goal exclude automation credential rotations as part of this first effort. Enabling the operator to be able to manually implement a credential rotation without disrupting a client's ability to function is, I think, a desirable outcome.

already implemented in the Supervisor.

#### Security Considerations

Some security considerations were already mentioned above. Here are a couple more.

##### Ensuring reasonable client secrets

During a login flow, when the client secret is presented in plaintext to the Supervisor’s token endpoint, it could
potentially validate that the secret meets some minimum entropy requirements. For example, it could check that the
secret has sufficient length, a sufficient number of unique characters, and a sufficient number of letter vs number
characters. If we choose to use the plaintext passwords option then the Supervisor could potentially perform this
validation in the mutating admission webhook before it hashes the passwords.

##### Disallowing audience confusion

Who decides the names of the dynamic client and the workload clusters?

- The name of the dynamic client would be chosen by the admin of the Supervisor. We could put validations on the name if
  we would like to limit the allowed names.
- The name of the workload cluster is chosen by the admin of the workload cluster (potentially a different person or

>   automated process). We don’t currently limit the string which chooses the audience name for the workload cluster, so

Reviewer 3: Question/thought: why do clusters have an audience name at all?  Why is this not simply the URL of the cluster?

Reviewer 3: Some related thoughts: a cluster could have multiple URLs, which is probably fine (sounds like something we can handle).  Uniqueness could be an issue, i.e. kind clusters can have the same URL, which would be a problem.  Maybe "audience" values that start with `https://` are special?

need to be regenerated to use the new client name (dynamic clients would not be allowed to use this reserved name).

###### Option 2.5: Some mix of Option 1 and 2

It is likely possible to use both options in conjunction.  Note that mixing old and new versions of the supervisor and
concierge may still result in audience confusion as components may not enforce all checks.

###### Option 3: Break apart signing keys for dynamic OAuth clients

Cryptographic isolation of signing keys is a guaranteed way to prevent any form of confusion between ID tokens issued
by the authorization code flow and and ID tokens issued by the token exchange API.  At a high level, we could separate
the single OIDC issuer into as many as four distinct issuers (authorization code vs. token exchange x static vs. dynamic
OAuth client).  Consumers would be configured to trust the exact sub-issuer that they are expected to understand.  The
OIDC discovery protocol makes this incredibly painful to express.  This technical specifics for this option have been
mostly omitted as it does not seem viable.

> ###### Option 4: Migrate to token webhook

Reviewer 3: An important bit that I forgot to include.

Two scenarios that we should aim to prevent:

1. The use of ID tokens from the auth code flow directly against the API server or used via the `JWTAuthenticator` (this is not a real problem currently because we only have the `pinniped-cli` today, but once we allow for dynamic clients, an admin could try to use dynamic clients to represent clusters - like they would with Dex - instead of using the token exchange API, or worse, they could try to re-use the dynamic clients across multiple clusters like they could with Dex)
2. Accidental use of the above by a "legacy" `JWTAuthenticator` (i.e. one configured for the existing pinniped TCR flow - this is partially described above as the audience confusion item `1` but is more referring to existing deployments of pinniped that are using the `JWTAuthenticator` today and, even without upgrading old workload clusters, should be retained as a safe configuration)

An approach to prevent the above while still allowing web apps to learn the identity of the user:

- Nest the `username` and `groups` claims in the auth code flow ID token under a new `pinniped` object and thus make it impossible for the `JWTAuthenticator`/Kube OIDC integration from consuming these ID tokens since that code lacks the ability to read into nested fields (i.e. force them to use the cluster scoped webhook tokens)
- Further gate the inclusion of the new `pinniped` object in the auth code flow ID token behind a new `profile` scope (this would not be requested by the CLI)
